### ISSUE TITLE
[BUG] Llama 3.2-1B exports for QCS6490 (Hexagon v68) produce degenerate output; genie_config rope-scaling factor wrong (8.0 vs 32.0)

### ISSUE THREAD

**Issue by webdevtodayjason** [qualcomm/ai-hub-models#337]
**Describe the bug**
Exporting `llama_v3_2_1b_instruct` with qai-hub-models 0.58.0 targeting `--chipset qualcomm-qcs6490` (Hexagon v68) completes successfully, but the resulting bundles produce degenerate output on-device (a token or two then EOS, or repetitive loops), for BOTH `--runtime genie` and `--runtime geniex_qairt`, with `--checkpoint DEFAULT_W4A16`.

Additionally, the generated `genie_config.json` contains `positional-encoding.rope-scaling.factor: 8.0`, while Llama 3.2's correct value is `32.0` (factor 8.0 is Llama 3.1's). The bundle's own copied HF `config.json` says 32.0, so the export pipeline is not propagating the model's rope_scaling. (Note: correcting the factor in genie_config does not fix the degenerate output — tested 8.0 / 32.0 / 1.0 — so there appears to be a second, deeper issue in the compiled context binaries for v68.)

**To Reproduce**
1. `pip install "qai-hub-models[llama-v3-2-1b-instruct]==0.58.0"` (Linux x86_64, aimet-onnx 2.35.1)
2. `python -m qai_hub_models.models.llama_v3_2_1b_instruct.export --checkpoint DEFAULT_W4A16 --runtime genie --chipset qualcomm-qcs6490 --skip-profiling --output-dir out` (also reproduced with `--runtime geniex_qairt`, and with/without `--context-lengths 1024`)
3. Deploy bundle to a QCS6490 device (Radxa Dragon Q6A, Ubuntu 24.04, QAIRT 2.46 runtime from the Ubuntu qcom PPA, DSP shells version-matched to firmware)
4. `genie-t2t-run -c genie_config.json --prompt_file prompt.txt`

Observed: `[BEGIN]:  On:[END]` (or degenerate repetition with other rope factors).

**Control test (rules out the device/runtime):** Radxa's older prebuilt Llama 3.2-1B bundle (compiled ~Aug 2025, QAIRT 2.40 era, single weight-sharing context binary, genie_config rope factor 32.0) produces coherent output at 12 tok/s on the SAME device, SAME QAIRT 2.46 runtime, SAME genie-t2t-run 2.46 binary.

**Compile/link details:** default `w4` precision fails at link time on this target with "Tensor 'embedding' has a floating-point type which is not supported by the targeted device" (expected on v68 — but perhaps w4 should be rejected earlier or auto-switched for fp16-less HTPs). `DEFAULT_W4A16` compiles and links successfully (QAIRT 2.45 tool_versions in metadata.json), it just doesn't produce a working model.

**Questions**
1. Is QCS6490/v68 still a supported LLM export target? The LLM-on-Genie tutorial now lists "Hexagon v73 or above", and the "QCS6490 (Proxy)" device appears to have been removed from AI Hub, yet `--chipset qualcomm-qcs6490` is accepted by the export scripts and compiles without warnings.
2. Is the rope-scaling factor 8.0 in generated genie_config.json for Llama 3.2 a known bug?

**Host configuration:**
- Export host: Ubuntu 24.04 x86_64, Python 3.12, qai-hub-models 0.58.0, qai-hub 0.53.0, aimet-onnx 2.35.1
- Target: Radxa Dragon Q6A (QCS6490, Hexagon v68), Ubuntu 24.04, kernel 6.18.2-4-qcom, QAIRT 2.46 (Ubuntu qcom PPA qairt-libs/qairt-tools/qairt-dsp-binaries), DSP firmware CDSP.HT.2.5.c4-00004-KODIAK-1 with matched fastrpc shells
- AI Hub job IDs available on request (compile SUCCESS, link SUCCESS for w4a16; link FAILED for w4 with the embedding fp error)

**Comment by webdevtodayjason** [comment]
Follow-up with a full per-graph isolation of the corruption, plus a link-stage bisect. **TL;DR: the defect is inside the compiled HTP graphs themselves — runtime, `genie_config`, rope tables, masks, and the cloud link stage are all exonerated.** The graphs produce correct next-token logits for the first ~2 prompt positions and then progressively collapse from position 3 onward. A Radxa 2.40-era bundle for the same model/board is correct at every position under the identical test harness.

## Method

Ran each split's graphs directly via `qnn-net-run` on the QCS6490 board (QAIRT 2.46 runtime), bypassing Genie entirely. All graph inputs (rope `position_ids_cos/sin`, causal `attention_mask`, zeroed KV at each tensor's quantization zero-point) were computed host-side from the HF reference model (fp32 `meta-llama/Llama-3.2-1B-Instruct`, llama3 rope scaling factor 32.0) and quantized bit-exactly to each graph's declared encodings. The identical harness was applied to the failing bundle (0.58.0 export, QAIRT 2.45 cloud compile) and to a known-good Radxa 2.40-era bundle as control.

## Results

**1. Embedding split is bit-perfect.** Six test tokens through `token_ar1_cl4096_1_of_3`: cosine 1.00000 vs HF `embed_tokens`, max abs error 5e-6 (= half the quant step).

**2. Prompt-graph (`ar128`) per-position logits progressively collapse — only in the new export.** Prompt: `The capital of France is Paris` (6 tokens, greedy argmax per position):

| pos | context so far | HF fp32 top-1 | Radxa 2.40 bundle | 0.58/2.45 export |
|---|---|---|---|---|
| 1 | The capital | ` of` | ` of` ✅ | ` of` ✅ |
| 2 | …capital of | ` the` | ` the` ✅ | ` the` ✅ |
| 3 | …of France | ` is` | ` is` ✅ | `,` — ` is` drops to 3rd ❌ |
| 4 | …France is | ` Paris` | ` Paris` ✅ | ` Paris` not in top-5 ❌ |
| 5 | …is Paris | `.` | `.` ✅ | `,` — distribution flattened ❌ |

Same rope tables, same mask construction, same KV zero-points, same harness code for both bundles. The control is correct everywhere; the new export degrades monotonically with position. This matches the on-device Genie symptom (prompt-fragment echoing, then repetitive degeneration).

**3. Corruption is in the COMPILE stage, not the cloud link stage.** I downloaded the six intermediate QNN DLCs (the compile-job outputs) and re-linked them into context binaries **locally on the board with QAIRT 2.46** (`qnn-context-binary-generator`), bypassing the AI Hub cloud link. The locally-linked `part2` produced **bit-identical** layer-8 hidden states to the cloud-linked bundle at every position (cosine 0.99964 at pos 0; 0.40 / 0.48 / 0.59 / 0.46 / 0.47 at pos 1–5 vs HF). Same DLCs through a different linker and QAIRT version = identical corruption ⇒ the defect is baked into the compiled DLCs (AIMET-ONNX quantize + HTP compile), not introduced at link. (Side note: on-board `qnn-context-binary-generator` reports `Weight Sharing is not supported on v68 target`, so the local relink used `weight_sharing_enabled=false`; this is a layout-only change and does not affect numerics.)

**4. `genie_config` rope-scaling factor (8.0 vs 32.0) is a separate, real-but-secondary bug.** Confirmed on-device: factor 8.0 → repetitive `On:` loop until context overflow; factor 32.0 → 2–3 tokens then early EOS. Neither is coherent, consistent with the graph-level corruption above dominating.

## Additional data points

- Failing compiles were produced by AI Hub cloud with QAIRT **2.45.0.260326154327** (`tool_versions.qairt` in bundle metadata), target `--target_runtime qnn_dlc --quantize_full_type w8a16 --quantize_io`, device chipset `qualcomm-qcs6490` (`hexagon:v68`).
- Reproduced across cl1024 and cl4096, and across `genie` and `geniex_qairt` runtimes.
- The known-good control is a Radxa 2.40-era bundle of the same model on the same silicon — so the regression is **specific to the newer QAIRT/AI-Hub compile path for v68**, not the model or the hardware.
- The 0.34.1 vintage recipe can no longer be used as a control through today's AI Hub: its compile options (`--quantize_full_type float16`) are rejected for v68 (`floating-point type not supported by the targeted device`), and `submit_link_job` now requires DLC inputs.

## Ask

Given the progressive positional corruption localized to graphs compiled by the current cloud QAIRT for v68 — bit-identical across two independent link stages, and absent from 2.40-era compiles of the same model on the same silicon — can you check for a v68-specific regression in the HTP graph compile/optimization path (plausibly attention/rope handling as attended context grows)? Happy to provide the full harness scripts, compile/link job IDs, and raw tensors.

**Comment by ashish-frozo** [comment]
The control test is the important part here: same device, same runtime, older prebuilt bundle works, new export compiles/links but produces bad output. That points to why a compile-success signal is not enough for release confidence.

I would keep the repro focused on three artifacts per run: generated genie_config, compile/link metadata, and a fixed prompt-output transcript from the target device. That makes it much easier to tell whether the failure is config propagation, graph compilation, or runtime execution.

**Comment by shreyajn** [comment]
A few questions:

Control test (rules out the device/runtime): Radxa's older prebuilt Llama 3.2-1B bundle (compiled ~Aug 2025, QAIRT 2.40 era, single weight-sharing context binary, genie_config rope factor 32.0) produces coherent output at 12 tok/s on the SAME device, SAME QAIRT 2.46 runtime, SAME genie-t2t-run 2.46 binary. -- Was this using AIHM?
